docs/getting-started/walkthroughs/datastream_api.md

```diff
@@ -0,0 +1,897 @@
+Either the current transaction caused a fraud alert, and the pattern is over, or the current transaction did not cause an alert, and the pattern is broken and needs to be restarted.
+
+Finally, the transaction amount is checked to see if it is small.
+If so, then the flag is set so that it can be checked by the next event.
+
+## State + Time = &#10084;&#65039; 
+
+Many event-driven applications require a strong notion of time to complement their state.
+For example, suppose you wanted to set a 24-hour timeout to your fraud detector; i.e., in the previous example transactions 3 and 4 would only be considered fraud if they occurred within 24 hours of each other.
+Flink allows setting timers which serve as callbacks at some point in time in the future.
+
+Along with the previous requirements:
+
+* Whenever the flag is set to true, also set a timer for 24 hours in the future.
+* When the timer fires, reset the flag by clearing its state.
+* If the flag is ever cleared the timer should be canceled.
+
+To cancel a timer, you have to remember what time it is set for, and remembering implies state, so you will begin by creating a timer state along with your flag state.
+
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+{% highlight java %}
+public class FraudDetector extends KeyedProcessFunction<Long, Transaction, Alert> {
+
+	public static final double SMALL_AMOUNT = 0.01;
+
+	public static final double LARGE_AMOUNT = 500.00;
+
+	public static final long ONE_DAY = 24 * 60 * 60 * 1000;
+
+	private transient ValueState<Boolean> flagState;
+
+	private transient ValueState<Long> timerState;
+
+	@Override
+	public void open(Configuration parameters) throws Exception {
+		ValueStateDescriptor<Boolean> flagDescriptor = new ValueStateDescriptor<>(
+				"flag",
+				Types.BOOLEAN);
+
+		flagState = getRuntimeContext().getState(flagDescriptor);
+
+		ValueStateDescriptor<Long> timerDescriptor = new ValueStateDescriptor<>(
+				"timer-state",
+				Types.LONG);
+
+		timerState = getRuntimeContext().getState(timerDescriptor);
+	}
+{% endhighlight %}
+</div>
+
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+class FraudDetector extends KeyedProcessFunction[Long, Transaction, Alert] {
+
+    @transient private var flagState: ValueState[java.lang.Boolean] = _
+
+    @transient private var timerState: ValueState[java.lang.Long] = _
+
+    @throws[Exception]
+    override def open(parameters: Configuration): Unit = {
+        val flagDescriptor = new ValueStateDescriptor("flag", Types.BOOLEAN)
+        flagState = getRuntimeContext.getState(flagDescriptor)
+
+        val timerDescriptor = new ValueStateDescriptor("timer-state", Types.LONG)
+        timerState = getRuntimeContext.getState(timerDescriptor)
+  }
+{% endhighlight %}
+</div>
+</div>
+
+`KeyedProcessFunction`'s are called with a `Context` that contains a timer service.
+The timer service can be used to query the current time, register timers, and delete timers.
```

Comment:
Since the code has `register**Processing**TimeTimer`, we should probably briefly mention what this is.